src/control/server/storage_scm.go

```diff
@@ -48,6 +67,166 @@ var (
 	msgScmUpdateNotImpl     = "scm firmware update not supported"
 )
 
+type runCmdFn func(string) ([]byte, error)
+
+// ScmSetup is an interface to configure scm prerequisites via shell tools
+type ScmSetup interface {
+	prep(common.ScmModules) ([]byte, error)
+	reset() error
+}
+
+// scmSetup is an implementation of the ScmSetup interface
+type scmSetup struct {
+	config *configuration
+}
+
+// prep executes commands to configure SCM modules into AppDirect
+// interleaved regions/sets hosting pmem kernel device namespaces.
+//
+// State will be established based on presence of modules, presence of
+// regions and free space on regions.
+//
+// Actions based on state:
+// * no modules -> no-op
+// * modules exist and no regions -> create all regions (needs reboot)
+// * regions exist and free capacity -> create all namespaces
+// * regions exist but no free capacity -> no-op
+//
+// Command output from external tools will be returned.
+func (s *scmSetup) prep(modules common.ScmModules) ([]byte, error) {
+	if len(modules) == 0 {
```

Comment:
Hmm. I'm not 100% certain without thinking about it some more, but this seems wrong. I agree that we should check for this condition and handle it, but I'm not sure I agree that it's not an error. It seems like a cleaner separation of concerns for the caller to decide whether or not to call this based on whether or not there are any modules. As it is currently, you're mixing concerns by making the caller check for a special string in the returned output, and this could lead to ambiguous behavior if the caller doesn't do what you expect.

Reply:
I've been pondering the exact same thing, moved it around a couple of times but haven't seen the correct place for it yet. I think it should be pushed up to the command layer tbh